### PR TITLE
pppoe: fix protocol field length variation

### PR DESCRIPTION
**Pull request by steffo**
Detect when protocol field is not a 16 bit field.
Added tests to prove logic
Updated existing test to use new test keywords

Ticket 4810

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [ X ] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [ X ] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/4810

Describe changes: 
Detect when protocol field is not a 16 bit field. 
Added tests to prove logic 
Updated existing test to use new test keywords

suricata-verify-pr: 582

### CODE REVIEW

File: src/decode-pppoe.c [hunk 1]
```diff
@@ -146,12 +145,31 @@ int DecodePPPOESession(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p,
...
+        if (ppp_protocol & 0x0100) {
+            /* Single-octet variant */
+            ppp_protocol >>= 8;
+            pppoesh_len = PPPOE_SESSION_HEADER_MIN_LEN;
+        } else {
+            /* Double-octet variant; increase the length of the session header accordingly */
+            pppoesh_len = PPPOE_SESSION_HEADER_MIN_LEN + 1;
+        }
+
+        if (len < pppoesh_len) {
```
Comment (catenacyber) [inline src/decode-pppoe.c:165]: Could we test this only if `pppoesh_len = PPPOE_SESSION_HEADER_MIN_LEN + 1;` ?

File: src/decode-pppoe.c [hunk 2]
```diff
@@ -185,46 +203,45 @@ int DecodePPPOESession(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p,
 
             case PPP_VJ_UCOMP:
 
-                if(len < (PPPOE_SESSION_HEADER_LEN + IPV4_HEADER_LEN))    {
+                if (len - pppoesh_len < IPV4_HEADER_LEN) {
```
Comment (catenacyber) [inline src/decode-pppoe.c:206]: Why did you move `pppoesh_len` to the other side of the comparison ?

**Review by catenacyber — COMMENTED** [review]
Thanks for this work.
I left some comments.

You could have split the api test changes, ie using `FAIL` into its own commit

File: src/decode-pppoe.c (hunk 1 above)
Comment (steffo) [reply src/decode-pppoe.c:165]: Yes, good catch

File: src/decode-pppoe.c (hunk 2 above)
Comment (steffo) [reply src/decode-pppoe.c:206]: It's to prevent a warning: comparison of integer expressions of different signedness: ‘uint32_t’

Comment (catenacyber) [reply src/decode-pppoe.c:206]: Ok.
I am not sure I understand the implicit integers conversions here, so I will trust the compiler

**Comment by steffo** [comment]
Replaced by #6780
